File: Vishwajeet/Day17.java
// Problem: Given a weighted fraph, use Djikstra's algorithm to find the shortest path and the distance between the given start and end nodes.

import java.util.*;

public class Day17 {
    private Map<Character, Map<Character, Integer>> adjList;

    // Constructor
    public Day17() {
        adjList = new HashMap<>();
    }

    public void addEdge(char from, char to, int weight) {
        adjList.putIfAbsent(from, new HashMap<>());
        adjList.get(from).put(to, weight);
        adjList.putIfAbsent(to, new HashMap<>()); // Ensure all nodes are initialized
    }

    // Djikstra's algorithm
    public Result djikstra(char start, char end) {
        Map<Character, Integer> distances = new HashMap<>();
        Map<Character, Character> previous = new HashMap<>();
        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(node -> node.distance));

        // Initialize distances
        for (char vertex : adjList.keySet()) {
            distances.put(vertex, Integer.MAX_VALUE);
        }
        distances.put(start, 0);
        pq.add(new Node(start, 0));

        while (!pq.isEmpty()) {
            Node currentNode = pq.poll();
            char current = currentNode.vertex;

            if (current == end) {
                break;
            }


            // Check if the distance to the current node is already greater than the distance to the current node
            for (Map.Entry<Character, Integer> neighborEntry : adjList.get(current).entrySet()) {
                char neighbor = neighborEntry.getKey();
                int newDist = distances.get(current) + neighborEntry.getValue();

                if (newDist < distances.get(neighbor)) {
                    distances.put(neighbor, newDist);
                    previous.put(neighbor, current);
                    pq.add(new Node(neighbor, newDist));
                }
            }
        }

        // Reconstruct path
        List<Character> path = new ArrayList<>();
        char step = end;
        if (previous.get(step) != null || step == start) {
            while (previous.get(step) != null) {
                path.add(step);
                step = previous.get(step);
            }
            path.add(start);
            Collections.reverse(path);
        }

        int distance = distances.get(end) == Integer.MAX_VALUE ? -1 : distances.get(end);
        return new Result(path, distance);
    }

    public static void main(String[] args) {
        Day17 graph = new Day17();

        // Add edges
        graph.addEdge('A', 'B', 4);
        graph.addEdge('A', 'C', 1);
        graph.addEdge('B', 'C', 2);
        graph.addEdge('B', 'D', 5);
        graph.addEdge('C', 'D', 8);
        graph.addEdge('C', 'E', 10);
        graph.addEdge('D', 'E', 2);

        char start = 'A';
        char end = 'E';
        Result result = graph.djikstra(start, end);

        if (result.distance == -1) {
            System.out.println("No path found from " + start + " to " + end);
        } else {
            System.out.println("Shortest path from " + start + " to " + end + ": " + result.path);
            System.out.println("Distance: " + result.distance);
        }
    }
}

class Node {
    char vertex;
    int distance;

    Node(char vertex, int distance) {
        this.vertex = vertex;
        this.distance = distance;
    }
}

class Result {
    List<Character> path;
    int distance;

    Result(List<Character> path, int distance) {
        this.path = path;
        this.distance = distance;
    }
}

File: Mohammad-Abbas/Day17.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <set>
#include <climits>
#include <algorithm>

using namespace std;

// Function to implement Dijkstra's algorithm using a set
void dijkstra(unordered_map<string, unordered_map<string, int>> &graph, string start, string end) {
    unordered_map<string, int> distances;
    unordered_map<string, string> previous;
    set<pair<int, string>> nodes;

    for (auto &node : graph) {
        distances[node.first] = INT_MAX;
    }
    distances[start] = 0;
    nodes.insert({0, start});

    while (!nodes.empty()) {
        auto top = *(nodes.begin());
        nodes.erase(nodes.begin());

        int current_distance = top.first;
        string current_node = top.second;

        for (auto &neighbor : graph[current_node]) {
            string next_node = neighbor.first;
            int edge_weight = neighbor.second;
            int new_distance = current_distance + edge_weight;

            if (new_distance < distances[next_node]) {
                if (distances[next_node] != INT_MAX) {
                    nodes.erase(nodes.find({distances[next_node], next_node}));
                }
                distances[next_node] = new_distance;
                previous[next_node] = current_node;
                nodes.insert({new_distance, next_node});
            }
        }
    }

    // Reconstruct the shortest path
    vector<string> path;
    for (string at = end; at != ""; at = previous[at]) {
        path.push_back(at);
    }
    reverse(path.begin(), path.end());

    // Print the shortest path and distance
    cout << "Shortest Path: ";
    for (const string &node : path) {
        cout << node << " ";
    }
    cout << "\nDistance: " << distances[end] << endl;
}

int main() {
    unordered_map<string, unordered_map<string, int>> graph = {
        {"A", {{"B", 4}, {"C", 1}}},
        {"B", {{"C", 2}, {"D", 5}}},
        {"C", {{"D", 8}, {"E", 10}}},
        {"D", {{"E", 2}}},
        {"E", {}}
    };

    string start = "A";
    string end = "E";

    dijkstra(graph, start, end);

    return 0;
}
